File: src/main/java/net/ropes4k/impl/FlatCharSequenceRope.java
/*
 * Copyright (C) 2024
 *  - Originally Copyright (C) 2007 Amin Ahmad.
 * Licenced under GPL
 */
package net.ropes4k.impl;

import net.ropes4k.Rope;

import java.io.IOException;
import java.io.Writer;
import java.util.Iterator;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * A rope constructed from an underlying character sequence.
 *
 * @author Amin Ahmad
 */
public final class FlatCharSequenceRope extends AbstractRope implements FlatRope {

    private final CharSequence sequence;

    /**
     * Constructs a new rope from an underlying character sequence.
     */
    public FlatCharSequenceRope(CharSequence sequence) {
        this.sequence = sequence;
    }

    @Override
    public char charAt(int index) {
        return sequence.charAt(index);
    }

    @Override
    public byte depth() {
        return 0;
    }

    @Override
    public Iterator<Character> iterator(int start) {
        if (start < 0 || start > length())
            throw new IndexOutOfBoundsException("Rope index out of range: " + start);
        return new Iterator<Character>() {
            int current = start;

            @Override
            public boolean hasNext() {
                return current < length();
            }

            @Override
            public Character next() {
                return sequence.charAt(current++);
            }

            @Override
            public void remove() {
                throw new UnsupportedOperationException("Rope iterator is read-only.");
            }
        };
    }

    @Override
    public int length() {
        return sequence.length();
    }

    @Override
    public Matcher matcher(Pattern pattern) {
        // optimized to return a matcher directly on the underlying sequence.
        return pattern.matcher(sequence);
    }

    @Override
    public Rope reverse() {
        return new ReverseRope(this);
    }

    @Override
    public Iterator<Character> reverseIterator(int start) {
        if (start < 0 || start > length())
            throw new IndexOutOfBoundsException("Rope index out of range: " + start);
        return new Iterator<Character>() {
            int current = length() - start;

            @Override
            public boolean hasNext() {
                return current > 0;
            }

            @Override
            public Character next() {
                return sequence.charAt(--current);
            }

            @Override
            public void remove() {
                throw new UnsupportedOperationException("Rope iterator is read-only.");
            }
        };
    }

    @Override
    public Rope subSequence(int start, int end) {
        if (start == 0 && end == length())
            return this;
        if (end - start < 8 || sequence instanceof String /* special optimization for String */) {
            return new FlatCharSequenceRope(sequence.subSequence(start, end));
        } else {
            return new SubstringRope(this, start, end - start);
        }
    }

    @Override
    public String toString() {
        return sequence.toString();
    }

    public String toString(int offset, int length) {
        return sequence.subSequence(offset, offset + length).toString();
    }

    @Override
    public void write(Writer out) throws IOException {
        write(out, 0, length());
    }

    @Override
    public void write(Writer out, int offset, int length) throws IOException {
        if (offset < 0 || offset + length > length())
            throw new IndexOutOfBoundsException("Rope index out of bounds:" + (offset < 0 ? offset : offset + length));

        if (sequence instanceof String) {    // optimization for String
            out.write(((String) sequence).substring(offset, offset + length));
            return;
        }
        for (int j = offset; j < offset + length; ++j)
            out.write(sequence.charAt(j));
    }
}

File: src/main/kotlin/net/ropes4k/impl/FlatCharSequenceRope.kt
/*
 * Copyright (C) 2024 James Richardson
 *  - Originally Copyright (C) 2007 Amin Ahmad.
 * Licenced under GPL
 */
package net.ropes4k.impl

import java.io.IOException
import java.io.Writer
import java.util.regex.Matcher
import java.util.regex.Pattern

/**
 * A rope constructed from an underlying character sequence.
 */
internal class FlatCharSequenceRope(private val chars: CharSequence) : AbstractRope(), FlatRope {

    override val length = chars.length

    override fun get(index: Int): Char {
        return chars[index]
    }

    override fun iterator(start: Int): Iterator<Char> {
        return FlatRopeIterator(this, start)
    }

    override fun matcher(pattern: Pattern): Matcher {
        // optimized to return a matcher directly on the underlying sequence.
        return pattern.matcher(chars)
    }

    override fun reverse(): InternalRope {
        return ReverseRope(this)
    }

    override fun reverseIterator(start: Int): Iterator<Char> {
        return FlatRopeReverseIterator(this, start)
    }

    override fun subSequence(startIndex: Int, endIndex: Int): InternalRope {
        if (startIndex == 0 && endIndex == length) return this
        return if (endIndex - startIndex < 8 || chars is String /* special optimization for String */) {
            FlatCharSequenceRope(chars.subSequence(startIndex, endIndex))
        } else {
            SubstringRope(this, startIndex, endIndex - startIndex)
        }
    }

    override fun toString(): String {
        return chars.toString()
    }

    override fun toString(offset: Int, length: Int): String {
        return chars.subSequence(offset, offset + length).toString()
    }

    @Throws(IOException::class)
    override fun write(out: Writer) {
        write(out, 0, length)
    }

    @Throws(IOException::class)
    override fun write(out: Writer, offset: Int, length: Int) {
        if (offset < 0 || offset + length > this.length) throw IndexOutOfBoundsException("Rope index out of bounds:" + (if (offset < 0) offset else offset + length))

        if (chars is String) {    // optimization for String
            out.write(chars.substring(offset, offset + length))
            return
        }
        for (j in offset until offset + length) out.write(chars[j].code)
    }
}
